### ISSUE TITLE
NetworkOnMainThread exception while resolving IP address.

### ISSUE THREAD

**Issue by emanueleDiVizio** [grpc/grpc-java#4154]
### What version of gRPC are you using?
1.9.0

### What did you expect to see?
When calling one of the stub's method I get a crash on Android, throwing NetworkOnMainThread exception with the following stacktrace. 

```java

02-16 22:52:08.352 10488-10488/com.ema.test W/ManagedChannelImpl: [io.grpc.internal.ManagedChannelImpl-1] Unexpected exception from LoadBalancer
                                                                      android.os.NetworkOnMainThreadException
                                                                          at android.os.StrictMode$AndroidBlockGuardPolicy.onNetwork(StrictMode.java:1448)
                                                                          at java.net.Inet6AddressImpl.getHostByAddr(Inet6AddressImpl.java:147)
                                                                          at java.net.InetAddress$1.getHostByAddr(InetAddress.java:266)
                                                                          at java.net.InetAddress.getHostFromNameService(InetAddress.java:569)
                                                                          at java.net.InetAddress.getHostName(InetAddress.java:515)
                                                                          at java.net.InetSocketAddress$InetSocketAddressHolder.getHostName(InetSocketAddress.java:83)
                                                                          at java.net.InetSocketAddress$InetSocketAddressHolder.-wrap2(Unknown Source:0)
                                                                          at java.net.InetSocketAddress.getHostName(InetSocketAddress.java:356)
                                                                          at io.grpc.internal.ProxyDetectorImpl.detectProxy(ProxyDetectorImpl.java:127)
                                                                          at io.grpc.internal.ProxyDetectorImpl.proxyFor(ProxyDetectorImpl.java:118)
                                                                          at io.grpc.internal.InternalSubchannel.startNewTransport(InternalSubchannel.java:209)
                                                                          at io.grpc.internal.InternalSubchannel.obtainActiveTransport(InternalSubchannel.java:190)
                                                                          at io.grpc.internal.ManagedChannelImpl$SubchannelImpl.requestConnection(ManagedChannelImpl.java:1020)
                                                                          at io.grpc.PickFirstBalancerFactory$PickFirstBalancer.handleResolvedAddressGroups(PickFirstBalancerFactory.java:79)
                                                                          at io.grpc.internal.ManagedChannelImpl$NameResolverListenerImpl$1NamesResolved.run(ManagedChannelImpl.java:924)
                                                                          at io.grpc.internal.ChannelExecutor.drain(ChannelExecutor.java:72)
                                                                          at io.grpc.internal.DelayedClientTransport.newStream(DelayedClientTransport.java:173)
                                                                          at io.grpc.internal.ClientCallImpl.start(ClientCallImpl.java:242)
                                                                          at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1.start(CensusTracingModule.java:387)
                                                                          at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1.start(CensusStatsModule.java:679)
                                                                          at io.grpc.stub.ClientCalls.startCall(ClientCalls.java:293)
                                                                          at io.grpc.stub.ClientCalls.asyncStreamingRequestCall(ClientCalls.java:283)
                                                                          at io.grpc.stub.ClientCalls.asyncBidiStreamingCall(ClientCalls.java:92)
                                                                          at session.SessionServiceGrpc$SessionServiceStub.startSession(SessionServiceGrpc.java:137)
```

My understanding is that the call `io.grpc.internal.ProxyDetectorImpl.detectProxy(ProxyDetectorImpl.java:127)` is happening in a `grpc-default-executor-0`, even though i've called `.executor(Executors.newSingleThreadExecutor())` on the channel builder. 
The `grpc-default-executor-0` thread pool appears to contain the main thread, as Android triggers a NetworkOnMainThread exception if any networking is attempted to be done on it.

It seems related to https://github.com/grpc/grpc-java/issues/577, despite it being marked as closed.

**Comment by ericgribkoff** [comment]
Thanks for reporting this. There is indeed a bug here, but the executor set on the channel builder is actually unrelated to this problem. `grpc-default-executor-0` is actually a thread itself, not a thread pool, and does not contain main or inadvertently run on the main Android UI thread.

Instead, what's happening is that the `ChannelExecutor` class you see in the stack trace is an internal gRPC class that handles running channel tasks serially, and it can and does often run light-weight tasks directly in the user thread. This is what is leading to the exception you're observing: when you start the gRPC call, it also drains the queued channel-related tasks, which in this instance includes (albeit inadvertently) looking up the proxy information. This triggers a network I/O event and the consequent `NetworkOnMainException`.

While we purposefully do some work on the calling thread, we aim to avoid any blocking network I/O operations. You're running into a bug in our proxy detection code here, as our invocation in [ProxyDetectorImpl](https://github.com/grpc/grpc-java/blob/3002a23a0ff9f72cc29c81b3be456d2ee1fa6bc4/core/src/main/java/io/grpc/internal/ProxyDetectorImpl.java#L127) of [`InetSocketAddress#getHostname()`](https://developer.android.com/reference/java/net/InetSocketAddress.html#getHostName()) can trigger a reverse lookup when the `InetSocketAddress` is created from a literal IP address. The quick-and-easy solution in the gRPC library for this particular problem is for us to use [`#getHostString()`](https://developer.android.com/reference/java/net/InetSocketAddress.html#getHostString()), which avoids the reverse lookup. However, this is only available on Android starting at API level 19+, so on earlier API levels we'd still have to fall back to `#getHostName()`. We already have a [`GrpcUtil#getHost()`](https://github.com/grpc/grpc-java/blob/3002a23a0ff9f72cc29c81b3be456d2ee1fa6bc4/core/src/main/java/io/grpc/internal/GrpcUtil.java#L579) method to perform this fallback, and it's likely just an oversight that it's not already being used in this instance.

If you're only targeting API levels >= 19, this quick fix should be sufficient. Let me know if you're willing to try patching in a snapshot version to help me verify that this change solves your problem (depending on your proxy environment, it may be possible we have other inadvertent network operations beyond this, so we'd want to make sure we're not missing any of them). I should be able to get this stop-gap into the next release, as well as backporting it to earlier versions - but it won't help you fully if you need to support older Android clients.

As far as addressing the root cause of this bug: there's an open PR (https://github.com/grpc/grpc-java/pull/4137) that centralizes all of the proxy detection work in the DNS resolver, which has its own separate executor, and will avoid doing any of this work on the calling thread.  cc @zpencer

In the interim, one "fix" is to work around this problem by avoiding invoking the gRPC calls in your main UI thread. Our Android examples use `AsyncTask` for this purpose, but there are other options as well. Depending on your application, this may be advisable anyways, as some gRPC related work (e.g., message serialization) might be relatively heavyweight even disregarding the network I/O.

**Comment by emanueleDiVizio** [comment]
Thanks @ericgribkoff, I've solved the issue by wrapping the grpc call in a separate thread. I don't have much time as of now to patch it myself, but let me know if you need any help.

**Comment by ericgribkoff** [comment]
Fixed by #4165
